pkg/jsonmessage/jsonmessage.go

```diff
@@ -151,69 +150,40 @@ type JSONMessage struct {
 	Aux *json.RawMessage `json:"aux,omitempty"`
 }
 
-/* Satisfied by gotty.TermInfo as well as noTermInfo from below */
-type termInfo interface {
-	Parse(attr string, params ...interface{}) (string, error)
+func clearLine(out io.Writer) {
+	eraseMode := aec.EraseModes.All
+	cl := aec.EraseLine(eraseMode)
+	fmt.Fprint(out, cl)
 }
 
-type noTermInfo struct{} // canary used when no terminfo.
-
-func (ti *noTermInfo) Parse(attr string, params ...interface{}) (string, error) {
-	return "", fmt.Errorf("noTermInfo")
-}
-
-func clearLine(out io.Writer, ti termInfo) {
-	// el2 (clear whole line) is not exposed by terminfo.
-
-	// First clear line from beginning to cursor
-	if attr, err := ti.Parse("el1"); err == nil {
-		fmt.Fprintf(out, "%s", attr)
-	} else {
-		fmt.Fprintf(out, "\x1b[1K")
-	}
-	// Then clear line from cursor to end
-	if attr, err := ti.Parse("el"); err == nil {
-		fmt.Fprintf(out, "%s", attr)
-	} else {
-		fmt.Fprintf(out, "\x1b[K")
-	}
-}
-
-func cursorUp(out io.Writer, ti termInfo, l int) {
+func cursorUp(out io.Writer, l int) {
 	if l == 0 { // Should never be the case, but be tolerant
 		return
 	}
-	if attr, err := ti.Parse("cuu", l); err == nil {
-		fmt.Fprintf(out, "%s", attr)
-	} else {
-		fmt.Fprintf(out, "\x1b[%dA", l)
-	}
+	up := aec.Up(uint(l))
```

Comment:
Looks at the callers I can quite convince myself that `l` is never negative, in one case it comes from `diff = len(ids) - line`, I suspect that might be always positive by construction (`line` is always derived from something like `len(ids)`), but I can't quite prove it is the case that `len(ids)` never shrinks..

I think this bug (if it is indeed a bug) was probably already there and I think would turn `Up(-1)` into effectively `Down(1)`, but this cast would make it `Up(INTMAX-1)` or something instead. Might be worth an explicit check? (perhaps in the caller with a change to the type of the arg to propagate the unsignedness?).

Not sure if this is really a blocker though.

Reply:
@ijc Do you think, just a simple check like `if l < 0` just `return` would work?

Reply:
Probably? Danger is that the existing code somehow relies on `cursorUp(-1)` moving down a line.

Picking through  `DisplayJSONMessagesStream` it seems we have:
```
				line = len(ids)
				ids[jm.ID] = line
```
so for any valid `X` the value of `ids[X]` is always <= `len(ids)` and so `diff = len(ids) - line` can never be negative. The `ids` map can be reset, but a `line = ids[jm.ID]` picked out of the old map cannot leak past that point by my reading (`diff` can, but that is safe since it is computed with the old, pre-reset, state).

So I think the check you suggest would work, but even better IMHO would be to try making `ids` be a `map[string]uint` not a `map[string]int` and adjust the types of `cursorUp` and `cursorDown` (and intermediate variables, most of which are autotyped) so that no cast is required in those functions. Only wrinkle is that `len()` returns an `int` so you would need the cast at the `len = uint(len(ids))` assignment, but there you know `len(ids)` cannot be negative (since a map cannot have negative size) and you can hope/assume that the len is not in the range of Go's INT_MAX→UINT_MAX (INT_MAX is huge).

Reply:
Thanks a lot for the reply @ijc 
I will push a commit with the suggested changes shortly.